core-bundle/src/Routing/Candidates/PageCandidates.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of Contao.
+ *
+ * (c) Leo Feyer
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Routing\Candidates;
+
+use Contao\CoreBundle\Routing\Page\PageRegistry;
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\FetchMode;
+use Doctrine\DBAL\Query\QueryBuilder;
+use Symfony\Component\HttpFoundation\Request;
+
+class PageCandidates extends AbstractCandidates
+{
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * @var PageRegistry
+     */
+    private $pageRegistry;
+
+    /**
+     * @var bool
+     */
+    private $initialized = false;
+
+    public function __construct(Connection $connection, PageRegistry $pageRegistry)
+    {
+        parent::__construct([], []);
+
+        $this->connection = $connection;
+        $this->pageRegistry = $pageRegistry;
+    }
+
+    public function getCandidates(Request $request): array
+    {
+        $this->initialize();
+
+        $candidates = parent::getCandidates($request);
+
+        $qb = $this->connection->createQueryBuilder();
+        $qb->select('id')->from('tl_page');
+
+        $hasRoot = $this->addRootQuery($candidates, $qb, $request->getHttpHost());
+        $hasRegex = $this->addRegexQuery($qb, $request->getPathInfo());
+
+        if ($hasRoot || $hasRegex) {
+            return array_unique(array_merge($candidates, $qb->execute()->fetchAll(FetchMode::COLUMN)));
+        }
+
+        return $candidates;
+    }
+
+    private function addRootQuery(array &$candidates, QueryBuilder $queryBuilder, string $httpHost): bool
+    {
+        if (!\in_array('index', $candidates, true)) {
+            return false;
+        }
+
+        $candidates[] = '/';
+
+        $queryBuilder->orWhere("type='root' AND (dns=:httpHost OR dns='')");
+        $queryBuilder->setParameter('httpHost', $httpHost);
+
+        return true;
+    }
+
+    private function addRegexQuery(QueryBuilder $queryBuilder, string $pathInfo): bool
+    {
+        $pathMap = $this->pageRegistry->getPathRegex();
+
+        if (empty($pathMap)) {
+            return false;
+        }
+
+        $paths = [];
+
+        foreach ($pathMap as $type => $pathRegex) {
+            $paths[] = '(?P<'.$type.'>'.substr($pathRegex, 2, strrpos($pathRegex, '$') - 2).')';
+        }
+
+        preg_match_all(
+            '#^(/'.implode('|/', array_map('preg_quote', $this->urlPrefixes)).')'.implode('|', $paths).'#sD',
```

Comment:
Can't we end on `$#sD` here? I once read that matching the beginning and start should always be done for performance reasons but maybe these days are long gone.

Reply:
@ausi /cc

Reply:
I don’t fully understand what the code does. Do we want to match a prefix or the complete path?

Reply:
I thought that would not be possible, because there is a variable URL suffix. However, as we do know the valid suffixes, we can add them the same as the prefixes. Implemented in 1f70bdbeacefb095fa911bd2930bbcdd9ed709fd and ded7d5c23918a0e5c90752bbaef6ccd17d149a99 😊

Reply:
+ imo `preg_quote` should always be called with the delimiter as second argument

Reply:
instead of the delimiter argument you could change the delimiter to `(` and `)`, these characters would get quoted anyways.

Reply:
> instead of the delimiter argument you could change the delimiter to `(` and `)`, these characters would get quoted anyways.

http://www.php.net/manual/en/function.preg-quote.php

> The special regular expression characters are: . \ + * ? [ ^ ] $ ( ) { } = ! < > | : - #

Reply:
Ah, didn’t know that `#` is a special character ☺️
Guess you’re the 🦊 now 🙃